access/ms-teams/app.go

```diff
@@ -0,0 +1,517 @@
+	}
+
+	a.mainJob.SetReady(ok)
+
+	<-a.watcherJob.Done()
+
+	return trace.Wrap(a.watcherJob.Err())
+}
+
+// checkTeleportVersion loads Teleport version and checks that it meets the minimal required
+func (a *App) checkTeleportVersion(ctx context.Context) (proto.PingResponse, error) {
+	log := logger.Get(ctx)
+	log.Debug("Checking Teleport server version")
+
+	pong, err := a.apiClient.WithCallOptions(grpc.WaitForReady(true)).Ping(ctx)
+	if err != nil {
+		if trace.IsNotImplemented(err) {
+			return pong, trace.Wrap(err, "server version must be at least %s", minServerVersion)
+		}
+
+		log.Error("Unable to get Teleport server version")
+		return pong, trace.Wrap(err)
+	}
+
+	err = lib.AssertServerVersion(pong, minServerVersion)
+
+	return pong, trace.Wrap(err)
+}
+
+// onWatcherEvent called when an access request event is received
+func (a *App) onWatcherEvent(ctx context.Context, event types.Event) error {
+	kind := event.Resource.GetKind()
+	if kind != types.KindAccessRequest {
+		return trace.Errorf("unexpected kind %s", kind)
+	}
+
+	op := event.Type
+	reqID := event.Resource.GetName()
+	ctx, _ = logger.WithField(ctx, "request_id", reqID)
+
+	switch op {
+	case types.OpPut:
+		ctx, _ = logger.WithField(ctx, "request_op", "put")
+		req, ok := event.Resource.(types.AccessRequest)
+		if !ok {
+			return trace.Errorf("unexpected resource type %T", event.Resource)
+		}
+		ctx, log := logger.WithField(ctx, "request_state", req.GetState().String())
+
+		var err error
+
+		switch {
+		case req.GetState().IsPending():
+			err = a.onPendingRequest(ctx, req)
+		case req.GetState().IsApproved():
+			err = a.onResolvedRequest(ctx, req)
+		case req.GetState().IsDenied():
+			err = a.onResolvedRequest(ctx, req)
+		default:
+			log.WithField("event", event).Warn("Unknown request state")
+			return nil
+		}
+
+		if err != nil {
+			log.WithError(err).Errorf("Failed to process request")
+			return trace.Wrap(err)
+		}
+
+		return nil
+
+	case types.OpDelete:
+		ctx, log := logger.WithField(ctx, "request_op", "delete")
+
+		if err := a.onDeletedRequest(ctx, reqID); err != nil {
+			log.WithError(err).Errorf("Failed to process deleted request")
+			return trace.Wrap(err)
+		}
+		return nil
+	default:
+		return trace.BadParameter("unexpected event operation %s", op)
+	}
+}
+
+// onPendingRequest is called when there's a new request/update to pending request
+func (a *App) onPendingRequest(ctx context.Context, req types.AccessRequest) error {
+	log := logger.Get(ctx)
+
+	id := req.GetName()
+	data := pd.AccessRequestData{
+		User:          req.GetUser(),
+		Roles:         req.GetRoles(),
+		RequestReason: req.GetRequestReason(),
+	}
+
+	pd, err := a.pd.Modify(ctx, id, func(existing interface{}) (interface{}, bool) {
+		if existing != nil {
+			return PluginData{}, false
+		}
+		return PluginData{AccessRequestData: data}, true
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// If we had not sent any messages before, let's create them
+	if pd != nil {
```

Comment:
Can we instead use `GetPluginData` to check if it exists already, instead of creating and updating the plugin data in two steps? I imagine this could lead to an out of sync state if the initial plugin data is created, but `postMessages` fails to update the app message or fails to update the plugin data.

Reply:
This would create a race condition if another instance of the plugin writes its data between get and modify calls.

Reply:
We should continue to use `CompareAndSwap` to avoid such race conditions, unless that's not the kind of race condition you're referring to?

The problem I'm pointing out is that when we create the initial `PluginData{AccessRequestData: data}` object, it is essentially useless until we add the `TeamsData` in the following `Modify` call. So if `a.bot.PostMessages` fails, this plugin data will remain in this useless state, and as I understand any future calls to update it would log `"Failed to post reply: plugin data is missing"`. If we somehow had another call to `onPendingRequest`, this would also fail after seeing the plugin data already exists, even though the messages were never actually posted.

Instead, after we first `GetPluginData` just to see if it already exists. If it does, we can do `a.bot.PostMessages`. If that succeeds, we can finally do `a.pd.Modify` (or `a.pd.Create` from my other comment`) to create the full plugin data `PluginData{AccessRequestData: data, TeamsData: teamsData}`. We can always be sure that if the plugin data exists, the messages have been posted. 

This still has the possible de-synced state of the messages being posted but the plugin data failing to be created, but I'm not sure there's a simple way to circumvent this.

Reply:
We can not get TeamsData unless we have messages actually sent. PluginData without TeamsData marks that an instance of the plugin has acquired a lock on this AccessRequest and is responsible for sending the initial messages. Yes, the situation when postMessages fails is possible. However, if we implement it in a way you explained, we could fall in the situation when two instances of the plugin determine that there is no plugin data and send duplicate messages.

Reply:
Ah ok that makes sense, thanks for explaining